### PR TITLE
feat(map): merge routes model and view model

### PR DESCRIPTION
**Pull request by MaelImhof**
# Summary

The map screen was created in three parts : the screen itself, the view model, and the repository. The view model and the repository had been worked on in parallel, hence one needed to put them together.

This pull request makes the map screen functional, in the sense that the user can now move around on the map and search for hikes wherever they want (see screenshots below).

To reach this objective :

- The view model `ListOfHikeRoutesViewModel.kt` was refactored to use `HikeRoute`s instead of `String`s
- The map screen was refactored to include a `Search here` button to avoid sending too many requests to the OSM API
- Tests were implemented and updated to cover the new behaviors

# Details

The high-level view of the changes is :

- The map now has a search button
- The user can move the map to where they want to search for hikes, then press the button
- Pressing the button will hide the button (to avoid spamming) and show a loading animation
- When the requests is finished (either succeeded or failed), the loading message disappears and the button is showed again
- If the request succeeded, found hikes are displayed in the list as before
- If the request failed, a toast is displayed to the user, indicating that there was an error

Also, all hike routes now come from the API. That is why their weird ID is displayed instead of a name, because the model does not provide a name yet.

## Side-changes

Apart from those main changes, two side-quests happened :

1. Added Mockito as an Android tests dependency (Mockito could not be used for UI testing with the previous setup, so I quickly modified `build.gradle.kts`, it works locally, I'm waiting to see the CI)
2. Removed hardcoded `Dispatchers.IO` from the view model and replaced them with an injected value, made Sonar happy and made the tests easier

## Additional features

Because this PR is merged after other PRs that impacted the map screen, it merges the previous PRs together with the added functionality. This results in a lot of new things starting to work together in this PR, maybe too many features for a single PR, but given the deadline we are facing I'm not sure opening several PRs and reviewing each of them would be a good idea.

The merged features are :

- The navigation PR that affected the map screen (#33)
- The PR for drawing hikes on the map (#48)

## Conclusion

Please let me know if you have any questions or any changes are required (most probably).

### CODE REVIEW

**Comment by MaelImhof** [comment]
There is one test that still needs fixing, I will see that tomorrow, for now I need some sleep. In any case, the rest of the PR is ready to be reviewed, there's only this test, `clickingOnHikeItemSelectsHikeRoute`, that needs more work.

By the way, any suggestion on how to fix it would be appreciated !

**Comment by MaelImhof** [comment]
Hope this fixes the problematic test.

Apart from the test that failed, I resolved conflicts with the `main` branch by integrating :

- The navigation PR that affected the map screen (#33)
- The PR for drawing hikes on the map (#48)

This makes the current PR contain a lot of new features, maybe too many for a single PR, unfortunately it is unavoidable given that this one is the last one to get merged, and given the deadlines we are facing.

**Review by mchac15 — CHANGES_REQUESTED** [review]
Overall I love the code. It is very well written and seems robust enough. Maybe read at the comments I made about your code.

File: app/src/main/java/ch/hikemate/app/model/route/ListOfHikeRoutesViewModel.kt [hunk 1]
```diff
@@ -1,53 +1,66 @@
...
-    // TODO: Should call the API repository to get all the routes filtered by the area
-    withContext(Dispatchers.IO) { Thread.sleep(100) }
-    hikeRoutes_.value = listOf("Route 1", "Route 2", "Route 3")
+  private suspend fun getRoutesAsync(onSuccess: () -> Unit = {}, onFailure: () -> Unit = {}) {
+    withContext(dispatcher) {
+      val area = area_.value ?: return@withContext
+      hikeRoutesRepository.getRoutes(
+          bounds = area.toBounds(),
+          onSuccess = { routes ->
+            hikeRoutes_.value = routes
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/model/route/ListOfHikeRoutesViewModel.kt:51]: That's the idea yes. `hikeRoutes_` is a `MutableStateFlow`. It contains a value, the "current" value, and can be observed by UI components (with `.collectAsState()`). When we say `hikeRoutes_.value = someValue`, it sends an update event to all the listeners that "subscribed" to this state flow.

In Jetpack Compose, update events will trigger a recomposition, leading to an update of the list of hikes in the UI.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt [hunk 2]
```diff
@@ -133,64 +163,48 @@ fun MapScreen(
...
+
+      // Maximum number of tiles that can be downloaded at once. Default is 40.
+      tileDownloadMaxQueueSize = 40
+
+      // Maximum number of bytes that can be used by the tile file system cache. Default is 600MB.
+      tileFileSystemCacheMaxBytes = 600L * 1024L * 1024L
+    }
+
+    mapView.apply {
+      controller.setZoom(12.0)
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:186]: Fair, I'll fix this real quick. I hope from the bottom of my heart that CI will still pass with new commits.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt [hunk 3]
```diff
@@ -133,64 +163,48 @@ fun MapScreen(
...
+    if (routes.size <= MapScreen.MAX_HIKES_DRAWN_ON_MAP) {
+      routes.forEach { showHikeOnMap(mapView, it, getRandomColor()) }
+      Log.d(MapScreen.LOG_TAG, "Displayed ${routes.size} hikes on the map")
+    } else {
+      Toast.makeText(
+              context,
+              context.getString(R.string.map_screen_too_many_hikes_message),
+              Toast.LENGTH_LONG)
+          .show()
+      Log.d(MapScreen.LOG_TAG, "Too many hikes (${routes.size}) to display on the map")
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:206]: Possible for sure, I did not think it made much sense when implementing it, but I added it in a commit I will push very soon.

**Review by mchac15 — APPROVED** [review]
LGTM

**Pull request merged**